From ca369aa30de37e78c0e9ecbf76d32ec0c01c46f9 Mon Sep 17 00:00:00 2001
From: Kevin Puetz <PuetzKevinA@JohnDeere.com>
Date: Fri, 25 Apr 2025 00:20:23 -0400
Subject: [PATCH] Implement the fix, remove todo_wine from the test

When CoMarshalInterface succeeds, it added an *additional* refcount
represented by the marshaling data written to the stream. The original
refcount returned from apartment_getclassobject needs to be released
regardless of marshaling success or failure.

Signed-off-by: Kevin Puetz <PuetzKevinA@JohnDeere.com>
---
 dlls/combase/apartment.c   | 3 +--
 dlls/ole32/tests/compobj.c | 2 +-
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/combase/apartment.c b/dlls/combase/apartment.c
index 85e06a6fd57..cb84731ebaf 100644
--- a/dlls/combase/apartment.c
+++ b/dlls/combase/apartment.c
@@ -1098,8 +1098,7 @@ static HRESULT apartment_hostobject(struct apartment *apt, const struct host_obj
         return hr;
 
     hr = CoMarshalInterface(params->stream, &params->iid, object, MSHCTX_INPROC, NULL, MSHLFLAGS_NORMAL);
-    if (FAILED(hr))
-        IUnknown_Release(object);
+    IUnknown_Release(object);
     IStream_Seek(params->stream, llZero, STREAM_SEEK_SET, NULL);
 
     return hr;
diff --git a/dlls/ole32/tests/compobj.c b/dlls/ole32/tests/compobj.c
index 5a2febdfc46..6ebbc6261d8 100644
--- a/dlls/ole32/tests/compobj.c
+++ b/dlls/ole32/tests/compobj.c
@@ -836,7 +836,7 @@ static void test_CoGetClassObject(void)
                 pUnk = NULL;
 
                 hr = (*pDllCanUnloadNow)();
-                todo_wine ok(S_OK == hr, "Unexpected hr %#lx.\n", hr);
+                ok(S_OK == hr, "Unexpected hr %#lx.\n", hr);
                 CloseHandle(htestlib);
             }
         }
-- 
GitLab

